File: Sources/MG101Core/PatchRecord.swift
import Foundation

public struct PatchRecord: Sendable {
    public private(set) var data: Data
    public let profile: DeviceProfile

    public init(data: Data, profile: DeviceProfile) throws {
        guard data.count == profile.recordSize else {
            throw PatchError.invalidSize(expected: profile.recordSize, actual: data.count)
        }
        self.data = data
        self.profile = profile
    }

    public init(contentsOf url: URL, profile: DeviceProfile) throws {
        try self.init(data: Data(contentsOf: url), profile: profile)
    }

    public var slotIndex: UInt32 {
        data.withUnsafeBytes { raw in
            raw.loadUnaligned(fromByteOffset: 0, as: UInt32.self).littleEndian
        }
    }

    public mutating func setSlotIndex(_ value: UInt32) {
        var littleEndian = value.littleEndian
        withUnsafeBytes(of: &littleEndian) {
            data.replaceSubrange(0..<4, with: $0)
        }
    }

    public var name: String {
        let range = profile.patchName.offset..<(profile.patchName.offset + profile.patchName.length)
        let bytes = data[range].prefix { $0 != 0 }
        return String(decoding: bytes, as: UTF8.self)
    }

    public var bpm: Int {
        Int(data[profile.bpm.msbOffset]) << 7 | Int(data[profile.bpm.lsbOffset])
    }

    public var irPresent: Bool {
        data[0x82] != 0 || data[0x83] != 0 || data[0x84] != 0 || data[0x85] != 0
    }

    public var irName: String {
        let bytes = data[0x86..<0xA6].prefix { $0 != 0 }
        return String(decoding: bytes, as: UTF8.self)
    }

    public func selector(for block: DeviceProfile.Block) -> UInt8 {
        data[block.selectorOffset]
    }

    public func modelID(for block: DeviceProfile.Block) -> Int {
        Int(selector(for: block) & 0x3F)
    }

    public func isBypassed(_ block: DeviceProfile.Block) -> Bool {
        selector(for: block) & 0x40 != 0
    }

    public func value(at offset: Int) -> Int {
        Int(data[offset])
    }

    public mutating func setByte(_ value: Int, at offset: Int) throws {
        guard data.indices.contains(offset) else { throw PatchError.offset(offset) }
        guard 0...255 ~= value else { throw PatchError.value(value) }
        data[offset] = UInt8(value)
    }

    public mutating func setBypass(_ bypassed: Bool, block: DeviceProfile.Block) {
        let model = data[block.selectorOffset] & 0x3F
        data[block.selectorOffset] = model | (bypassed ? 0x40 : 0)
    }

    public mutating func setParameter(
        _ parameter: EffectCatalog.Parameter,
        value: Int
    ) throws {
        guard parameter.minimum...parameter.maximum ~= value else {
            throw PatchError.parameterRange(
                parameter.name,
                parameter.minimum,
                parameter.maximum,
                value
            )
        }
        try setByte(value, at: parameter.fileOffset)
    }

    public mutating func setModel(
        _ model: EffectCatalog.Model,
        block: DeviceProfile.Block,
        values: [Int],
        bypassed: Bool,
        clearInactive: Bool = true
    ) throws {
        guard values.count == model.parameters.count else {
            throw PatchError.parameterCount(model.parameters.count, values.count)
        }
        data[block.selectorOffset] = UInt8(model.modelID) | (bypassed ? 0x40 : 0)
        let active = Set(model.parameters.map(\.fileOffset))
        for (parameter, value) in zip(model.parameters, values) {
            try setParameter(parameter, value: value)
        }
        if clearInactive {
            for offset in block.parameterOffsets where !active.contains(offset) {
                data[offset] = 0
            }
        }
    }

    public mutating func setBPM(_ value: Int) throws {
        guard profile.bpm.minimum...profile.bpm.maximum ~= value else {
            throw PatchError.bpm(value)
        }
        data[profile.bpm.msbOffset] = UInt8((value >> 7) & 0x7F)
        data[profile.bpm.lsbOffset] = UInt8(value & 0x7F)
    }

    public mutating func setNamedField(_ name: String, value: Int) throws {
        guard let field = profile.namedFields[name] else {
            throw PatchError.unknownField(name)
        }
        guard field.minimum...field.maximum ~= value else {
            throw PatchError.parameterRange(
                name,
                field.minimum,
                field.maximum,
                value
            )
        }
        try setByte(value, at: field.offset)
    }

    public mutating func setName(_ value: String) throws {
        let encoded = Array(value.utf8)
        guard encoded.count <= profile.patchName.length else {
            throw PatchError.nameTooLong(profile.patchName.length)
        }
        let start = profile.patchName.offset
        for index in 0..<profile.patchName.length {
            data[start + index] = index < encoded.count ? encoded[index] : 0
        }
    }

    public mutating func setIR(wav: Data, name: String) throws {
        let expectedLength = profile.recordSize - 0xA6
        guard wav.count == expectedLength else {
            throw PatchError.ir("Expected \(expectedLength) WAV bytes, received \(wav.count).")
        }
        guard wav.prefix(4) == Data("RIFF".utf8),
              wav[8..<12] == Data("WAVE".utf8)
        else {
            throw PatchError.ir("IR must be a RIFF/WAVE file.")
        }
        let riffSize = wav.withUnsafeBytes {
            $0.loadUnaligned(fromByteOffset: 4, as: UInt32.self).littleEndian
        }
        guard riffSize == UInt32(expectedLength - 8) else {
            throw PatchError.ir("Invalid RIFF size \(riffSize).")
        }
        let encoded = Array(name.utf8)
        guard encoded.count <= 32 else { throw PatchError.ir("IR name exceeds 32 bytes.") }

        data[0x82] = 1
        data[0x83] = 0
        data[0x84] = 0
        data[0x85] = 0
        for index in 0..<32 {
            data[0x86 + index] = index < encoded.count ? encoded[index] : 0
        }
        data.replaceSubrange(0xA6..<profile.recordSize, with: wav)
    }

    public mutating func clearIR() {
        data.replaceSubrange(0x82..<profile.recordSize, with: repeatElement(0, count: profile.recordSize - 0x82))
    }

    public func differences(from original: PatchRecord) -> [ByteDifference] {
        zip(original.data, data).enumerated().compactMap { offset, pair in
            pair.0 == pair.1 ? nil : ByteDifference(
                offset: offset,
                before: pair.0,
                after: pair.1
            )
        }
    }
}

public struct ByteDifference: Identifiable, Hashable, Sendable {
    public let offset: Int
    public let before: UInt8
    public let after: UInt8
    public var id: Int { offset }
}

public enum PatchError: LocalizedError {
    case invalidSize(expected: Int, actual: Int)
    case offset(Int)
    case value(Int)
    case bpm(Int)
    case nameTooLong(Int)
    case parameterRange(String, Int, Int, Int)
    case parameterCount(Int, Int)
    case unknownField(String)
    case ir(String)

    public var errorDescription: String? {
        switch self {
        case .invalidSize(let expected, let actual):
            "Expected \(expected) bytes, received \(actual)."
        case .offset(let value): "Offset outside record: \(value)."
        case .value(let value): "Byte value outside 0...255: \(value)."
        case .bpm(let value): "BPM outside configured range: \(value)."
        case .nameTooLong(let maximum): "Patch name exceeds \(maximum) UTF-8 bytes."
        case .parameterRange(let name, let minimum, let maximum, let value):
            "\(name)=\(value) is outside \(minimum)...\(maximum)."
        case .parameterCount(let expected, let actual):
            "Expected \(expected) parameter values, received \(actual)."
        case .unknownField(let name): "Unknown named field: \(name)."
        case .ir(let message): "Invalid local IR: \(message)"
        }
    }
}

File: rust/crates/core/src/patch_record.rs
//! Rekord patcha — wierny port `PatchRecord` ze Swift (MG101Core).
//!
//! Rekord to bufor bajtów o rozmiarze `record_size` z akcesorami sterowanymi
//! profilem. „Bajty święte" (ADR-0002): nieznane bajty nietknięte przy edycji.
//! Offsety regionu IR (`0x82`/`0x86`/`0xA6`) są — jak w Swift — stałymi MG-101.

use crate::device_profile::{Block, DeviceProfile};
use crate::effect_catalog::{Model, Parameter};
use crate::error::PatchError;

const IR_FLAG_OFFSET: usize = 0x82;
const IR_NAME_OFFSET: usize = 0x86;
const IR_NAME_LEN: usize = 32;
const IR_WAV_OFFSET: usize = 0xA6;

/// Różnica pojedynczego bajtu (offset + wartość przed/po).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ByteDifference {
    pub offset: usize,
    pub before: u8,
    pub after: u8,
}

/// Rekord patcha powiązany z profilem urządzenia.
#[derive(Debug, Clone)]
pub struct PatchRecord<'p> {
    data: Vec<u8>,
    profile: &'p DeviceProfile,
}

impl<'p> PatchRecord<'p> {
    /// Tworzy rekord, sprawdzając rozmiar względem profilu.
    pub fn new(data: Vec<u8>, profile: &'p DeviceProfile) -> Result<Self, PatchError> {
        if data.len() != profile.record_size {
            return Err(PatchError::InvalidSize {
                expected: profile.record_size,
                actual: data.len(),
            });
        }
        Ok(Self { data, profile })
    }

    /// Surowe bajty rekordu.
    pub fn data(&self) -> &[u8] {
        &self.data
    }

    /// Profil powiązany.
    pub fn profile(&self) -> &DeviceProfile {
        self.profile
    }

    /// Indeks slotu (UInt32 LE, offset 0).
    pub fn slot_index(&self) -> u32 {
        u32::from_le_bytes([self.data[0], self.data[1], self.data[2], self.data[3]])
    }

    /// Ustawia indeks slotu (LE).
    pub fn set_slot_index(&mut self, value: u32) {
        self.data[0..4].copy_from_slice(&value.to_le_bytes());
    }

    /// Nazwa patcha (UTF-8, ucięta na pierwszym `0`).
    pub fn name(&self) -> String {
        let start = self.profile.patch_name.offset;
        let end = start + self.profile.patch_name.length;
        let slice = &self.data[start..end];
        let trimmed: Vec<u8> = slice.iter().copied().take_while(|&b| b != 0).collect();
        String::from_utf8_lossy(&trimmed).into_owned()
    }

    /// BPM = msb<<7 | lsb.
    pub fn bpm(&self) -> i64 {
        (i64::from(self.data[self.profile.bpm.msb_offset]) << 7)
            | i64::from(self.data[self.profile.bpm.lsb_offset])
    }

    /// Czy obecny jest lokalny IR (któryś z bajtów flagi `0x82..=0x85` ≠ 0).
    pub fn ir_present(&self) -> bool {
        self.data[0x82] != 0 || self.data[0x83] != 0 || self.data[0x84] != 0 || self.data[0x85] != 0
    }

    /// Nazwa IR (UTF-8, ucięta na `0`).
    pub fn ir_name(&self) -> String {
        let slice = &self.data[IR_NAME_OFFSET..IR_NAME_OFFSET + IR_NAME_LEN];
        let trimmed: Vec<u8> = slice.iter().copied().take_while(|&b| b != 0).collect();
        String::from_utf8_lossy(&trimmed).into_owned()
    }

    /// Bajt selektora bloku.
    pub fn selector(&self, block: &Block) -> u8 {
        self.data[block.selector_offset]
    }

    /// ID modelu (dolne 6 bitów selektora).
    pub fn model_id(&self, block: &Block) -> i64 {
        i64::from(self.selector(block) & 0x3F)
    }

    /// Czy blok jest zbypassowany (bit `0x40`).
    pub fn is_bypassed(&self, block: &Block) -> bool {
        self.selector(block) & 0x40 != 0
    }

    /// Wartość bajtu pod offsetem.
    pub fn value_at(&self, offset: usize) -> i64 {
        i64::from(self.data[offset])
    }

    /// Ustawia bajt (0..=255) pod offsetem, z kontrolą zakresu i granic.
    pub fn set_byte(&mut self, value: i64, offset: usize) -> Result<(), PatchError> {
        if offset >= self.data.len() {
            return Err(PatchError::Offset(offset));
        }
        if !(0..=255).contains(&value) {
            return Err(PatchError::Value(value));
        }
        self.data[offset] = value as u8;
        Ok(())
    }

    /// Ustawia/zdejmuje bypass bloku, zachowując id modelu.
    pub fn set_bypass(&mut self, bypassed: bool, block: &Block) {
        let model = self.data[block.selector_offset] & 0x3F;
        self.data[block.selector_offset] = model | if bypassed { 0x40 } else { 0 };
    }

    /// Ustawia parametr (z kontrolą zakresu z katalogu).
    pub fn set_parameter(&mut self, parameter: &Parameter, value: i64) -> Result<(), PatchError> {
        if !(parameter.minimum()..=parameter.maximum()).contains(&value) {
            return Err(PatchError::ParameterRange {
                name: parameter.name.clone(),
                minimum: parameter.minimum(),
                maximum: parameter.maximum(),
                value,
            });
        }
        self.set_byte(value, parameter.file_offset)
    }

    /// Ustawia model bloku: selektor + wszystkie parametry; opcjonalnie zeruje
    /// offsety parametrów nieaktywne w nowym modelu (wierny port `setModel`).
    pub fn set_model(
        &mut self,
        model: &Model,
        block: &Block,
        values: &[i64],
        bypassed: bool,
        clear_inactive: bool,
    ) -> Result<(), PatchError> {
        if values.len() != model.parameters.len() {
            return Err(PatchError::ParameterCount {
                expected: model.parameters.len(),
                actual: values.len(),
            });
        }
        self.data[block.selector_offset] = (model.model_id as u8) | if bypassed { 0x40 } else { 0 };
        let active: std::collections::HashSet<usize> =
            model.parameters.iter().map(|p| p.file_offset).collect();
        for (parameter, &value) in model.parameters.iter().zip(values.iter()) {
            self.set_parameter(parameter, value)?;
        }
        if clear_inactive {
            for &offset in &block.parameter_offsets {
                if !active.contains(&offset) {
                    self.data[offset] = 0;
                }
            }
        }
        Ok(())
    }

    /// Ustawia BPM (rozbicie na 7-bitowe MSB/LSB).
    pub fn set_bpm(&mut self, value: i64) -> Result<(), PatchError> {
        if !(self.profile.bpm.minimum..=self.profile.bpm.maximum).contains(&value) {
            return Err(PatchError::Bpm(value));
        }
        self.data[self.profile.bpm.msb_offset] = ((value >> 7) & 0x7F) as u8;
        self.data[self.profile.bpm.lsb_offset] = (value & 0x7F) as u8;
        Ok(())
    }

    /// Ustawia pole globalne (send/return/patch.*).
    pub fn set_named_field(&mut self, name: &str, value: i64) -> Result<(), PatchError> {
        let field = self
            .profile
            .named_fields
            .get(name)
            .ok_or_else(|| PatchError::UnknownField(name.to_string()))?;
        if !(field.minimum..=field.maximum).contains(&value) {
            return Err(PatchError::ParameterRange {
                name: name.to_string(),
                minimum: field.minimum,
                maximum: field.maximum,
                value,
            });
        }
        let offset = field.offset;
        self.set_byte(value, offset)
    }

    /// Ustawia nazwę patcha (UTF-8, dopełnienie zerami do długości pola).
    pub fn set_name(&mut self, value: &str) -> Result<(), PatchError> {
        let encoded = value.as_bytes();
        let len = self.profile.patch_name.length;
        if encoded.len() > len {
            return Err(PatchError::NameTooLong(len));
        }
        let start = self.profile.patch_name.offset;
        for index in 0..len {
            self.data[start + index] = encoded.get(index).copied().unwrap_or(0);
        }
        Ok(())
    }

    /// Osadza IR (WAV) — wierny port `setIR` z walidacją RIFF/WAVE i długości.
    pub fn set_ir(&mut self, wav: &[u8], name: &str) -> Result<(), PatchError> {
        let expected_length = self.profile.record_size - IR_WAV_OFFSET;
        if wav.len() != expected_length {
            return Err(PatchError::Ir(format!(
                "Expected {expected_length} WAV bytes, received {}.",
                wav.len()
            )));
        }
        if &wav[0..4] != b"RIFF" || &wav[8..12] != b"WAVE" {
            return Err(PatchError::Ir("IR must be a RIFF/WAVE file.".into()));
        }
        let riff_size = u32::from_le_bytes([wav[4], wav[5], wav[6], wav[7]]);
        if riff_size != (expected_length - 8) as u32 {
            return Err(PatchError::Ir(format!("Invalid RIFF size {riff_size}.")));
        }
        let encoded = name.as_bytes();
        if encoded.len() > IR_NAME_LEN {
            return Err(PatchError::Ir("IR name exceeds 32 bytes.".into()));
        }
        self.data[IR_FLAG_OFFSET] = 1;
        self.data[0x83] = 0;
        self.data[0x84] = 0;
        self.data[0x85] = 0;
        for index in 0..IR_NAME_LEN {
            self.data[IR_NAME_OFFSET + index] = encoded.get(index).copied().unwrap_or(0);
        }
        self.data[IR_WAV_OFFSET..self.profile.record_size].copy_from_slice(wav);
        Ok(())
    }

    /// Usuwa IR (zeruje region `0x82..record_size`).
    pub fn clear_ir(&mut self) {
        for b in &mut self.data[IR_FLAG_OFFSET..self.profile.record_size] {
            *b = 0;
        }
    }

    /// Różnice bajtowe względem oryginału (offset/before/after) — port `differences`.
    pub fn differences(&self, original: &PatchRecord) -> Vec<ByteDifference> {
        original
            .data
            .iter()
            .zip(self.data.iter())
            .enumerate()
            .filter_map(|(offset, (&before, &after))| {
                if before == after {
                    None
                } else {
                    Some(ByteDifference {
                        offset,
                        before,
                        after,
                    })
                }
            })
            .collect()
    }
}
